src/mlpack/tests/main_tests/approx_kfn_test.cpp

```diff
@@ -0,0 +1,251 @@
+ */
+BOOST_AUTO_TEST_CASE(ApproxKFNZeroNumTablesTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(2, 80); // 80 points in 2 dimensions.
+
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("num_tables", (int) 0); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Check that we can't specify num_tables as negative.
+ */
+BOOST_AUTO_TEST_CASE(ApproxKFNNegativeNumTablesTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(2, 80); // 80 points in 2 dimensions.
+ 
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("num_tables", (int) -5); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Ensuring that a saved model can be loaded and used again correctly.
+ */
+BOOST_AUTO_TEST_CASE(ApproxKFNModelReuseTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(2, 80); // 80 points in 2 dimensions.
+
+  arma::mat queryData;
+  queryData.randu(2, 40); // 40 points in 2 dimensions.
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("query", queryData);
+  SetInputParam("k", (int) 10);
+
+  mlpackMain();
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+  neighbors = std::move(CLI::GetParam<arma::Mat<size_t>>("neighbors"));
+  distances = std::move(CLI::GetParam<arma::mat>("distances"));
+
+  // Reset passed parameters.
+  CLI::GetSingleton().Parameters()["reference"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["query"].wasPassed = false;
+
+  // Input saved model, pass the same query and keep k unchanged.
+  SetInputParam("input_model",
+      std::move(CLI::GetParam<ApproxKFNModel*>("output_model")));
+  SetInputParam("query", queryData);
+
+  mlpackMain();
+
+  // Check that initial output matrices and the output matrices using
+  // saved model are equal.
+  CheckMatrices(neighbors, CLI::GetParam<arma::Mat<size_t>>("neighbors"));
+  CheckMatrices(distances, CLI::GetParam<arma::mat>("distances"));
+}
+
+/**
+ * Make sure that the dimensions of the exact distances matrix are correct.
+ */
+BOOST_AUTO_TEST_CASE(ApproxKFNExactDistDimensionTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(2, 80); // 80 points in 2 dimensions.
+
+  SetInputParam("reference", std::move(referenceData));
+  // Random input, any k <= reference points.
+  SetInputParam("k", (int) 10);
+  SetInputParam("calculate_error", (bool) true);
+
+  // Random matrix specifying exact distances of each point to its k neighbors.
+  // Note that the values in the matrix do not matter as we are only concernec
+  // with the dimensions of the matrix passed.
+  arma::mat exactDistances;
+  exactDistances.randu(9, 90); // Wrong size (should be (10, 80)).
```

Comment:
I think the size here should actually be `(1, 80)`, but in any case, the test still works, I am only saying that because I think the comment is incorrect.

Reply:
Oh, I might have understood exact distances incorrectly. Doesn't the matrix contain values for each of the k furthest neighbors for each point in the reference set?

Reply:
Hmm, I guess the program actually doesn't really care how many rows are in the `exact_distances` matrix as long as it is more than one, so I suppose it doesn't matter.  Still, I think `(1, 80)` might be a better comment in case anyone reads this test trying to figure out how to use the binding, since only that first row is needed.